### DISCUSSION TITLE
Can't run handleCreatePaymentRedirect method

### DISCUSSION THREAD

**Discussion by cusursuz** [impress-org/givewp#7218] [Developer Q&A]
I implemented custom gateway conform
https://givewp.com/documentation/developers/how-to-build-a-gateway-add-on-for-givewp/

can't run handleCreatePaymentRedirect method after receiving data from bank. Signature is valid and has not expired
How can run it?
thanks

`<php
require_once(__DIR__ . '/vendor/autoload.php');

use Give\Donations\Models\Donation;
use Give\Donations\Models\DonationNote;
use Give\Donations\ValueObjects\DonationStatus;
use Give\Framework\Exceptions\Primitives\Exception;
use Give\Framework\Http\Response\Types\RedirectResponse;
use Give\Framework\PaymentGateways\Commands\PaymentRefunded;
use Give\Framework\PaymentGateways\Commands\RedirectOffsite;
use Give\Framework\PaymentGateways\PaymentGateway;

use Monolog\Handler\StreamHandler;
use Monolog\Logger;
use GuzzleHttp\Client;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\MessageFormatter;
use Maib\MaibApi\MaibClient;

/**
 * @inheritDoc
 */
class ExampleGatewfayOffsiteClass extends PaymentGateway
{
    private $client;

    public function __construct()
    {
        $options = [
            'base_uri' => MaibClient::MAIB_TEST_BASE_URI,
            'debug'  => false,
            'verify' => true,
            'cert'    => [MaibClient::MAIB_TEST_CERT_URL, MaibClient::MAIB_TEST_CERT_PASS],
            'ssl_key' => MaibClient::MAIB_TEST_CERT_KEY_URL,
            'config'  => [
              'curl'  =>  [
                CURLOPT_SSL_VERIFYHOST => 2,
                CURLOPT_SSL_VERIFYPEER => true,
              ]
            ]
          ];

        $guzzleClient = new Client($options);
        $this->client = new MaibClient($guzzleClient);
    }
    /**
     * @inheritDoc
     */
    public $secureRouteMethods = [
        'handleCreatePaymentRedirect',
    ];

    /**
     * @inheritDoc
     */
    public static function id(): string
    {
        return 'maib-gateway-offsite';
    }

    /**
     * @inheritDoc
     */
    public function getId(): string
    {
        return self::id();
    }

    /**
     * @inheritDoc
     */
    public function getName(): string
    {
        return __('MAIB Gateway - Offsite', 'maib-give');
    }

    /**
     * @inheritDoc
     */
    public function getPaymentMethodLabel(): string
    {
        return __('MAIB Gateway - Offsite', 'maib-give');
    }

    /**
     * @inheritDoc
     */
    public function getLegacyFormFieldMarkup(int $formId, array $args): string
    {
        // For an offsite gateway, this is just help text that displays on the form. 
        return "<div class='example-offsite-help-text'>
                    <p>You will be taken away to MAIB to complete the donation!</p>
                </div>";
    }

    /**
     * Get Sample Data to send to a gateway.
     */
    public function getParameters(Donation $donation): array
    {
        // this is just an example but,
        // you would need to get your own merchant ID and key from the gateway.
        // Typically, these are retrieved from the GiveWP gateway settings when the admin
        // sets up the gateway using their account.

        return [
            'cancel_url' => give_get_failed_transaction_uri(),
            'notify_url' => get_site_url() . '/?give-listener=MAIB',
            'name_first' => $donation->firstName,
            'name_last' => $donation->lastName,
            'email_address' => $donation->email,
            'm_payment_id' => $donation->id,
            'amount' => $donation->amount->formatToDecimal(),
            'item_name' => $donation->formTitle,
            'item_description' => sprintf(__('Donation via GiveWP, ID %s', 'maib-give'), $donation->id),
        ];
    }

    /**
     * @inheritDoc
     */
    public function createPayment(Donation $donation, $gatewayData)
    {   
        // The Parameters required to use MaibClient methods
        $amount = $donation->amount->formatToDecimal(); // The amount of the transaction

        //MDL $donation->amount->getCurrency()->getCode();

        $currency = 498; // The currency of the transaction - is the 3 digits code of currency from ISO 4217
        $clientIpAddr = $donation->donorIp; // The client IP address
        $description = 'Donation';// . $donation->currency; // The description of the transaction
        $lang = 'en'; // The language for the payment gateway

        // Other parameters
        $redirect_url = MaibClient::MAIB_TEST_REDIRECT_URL . '?trans_id=';

        $registerSmsTransaction = $this->client->registerSmsTransaction($amount, $currency, $clientIpAddr, $description, $lang);
        $sms_transaction_id = $registerSmsTransaction["TRANSACTION_ID"];
        $sms_redirect_url = $redirect_url . $sms_transaction_id;

        $donation->gatewayTransactionId = $sms_transaction_id;
        $donation->save();

        // Step 1: generate a secure gateway route URL that will be used to redirect the donor to the gateway.
        // The args you pass it will be available in the $queryParams parameter in the secureRouteMethod that you defined below.
        $returnUrl = $this->generateSecureGatewayRouteUrl(
            'handleCreatePaymentRedirect',
            $donation->id,
            [
                'givewp-donation-id' => $donation->id,
                'givewp-success-url' => urlencode(give_get_success_page_uri()),
                // this would likely be a transaction ID from the gateway upon return.
                'givewp-gateway-transaction-id' => urlencode($sms_transaction_id),//'123456789',
            ]
        );

        // Step 2: Get the parameters you need to send to the gateway.
        $queryParams = array_merge(
            $this->getParameters($donation),
            // this just an example of a return url parameter, this will be specific to your gateway.
            ['return_url' => $returnUrl, 'transaction_id' => $sms_transaction_id]
        );

        // Step 3: Generate the URL to redirect the donor to, using the queryParams you created that contains the secure gateway route URL.
        $gatewayUrl = add_query_arg($queryParams, $sms_redirect_url);

        // Step 4: Return a RedirectOffsite command with the generated URL to redirect the donor to the gateway.
        return new RedirectOffsite($gatewayUrl);
    }

    /**
     * An example of using a secureRouteMethod for extending the Gateway API to handle a redirect.
     *
     * @throws Exception
     */
    protected function handleCreatePaymentRedirect(array $queryParams): RedirectResponse
    {
        // Step 1: Use the $queryParams to get the data you need to complete the donation.
        $donationId = $queryParams['givewp-donation-id'];
        $gatewayTransactionId = $queryParams['givewp-gateway-transaction-id'];
        // $successUrl = $queryParams['givewp-success-url'];

        // Step 2: Typically you will find the donation from the donation ID.
        /** @var Donation $donation */
        $donation = Donation::find($donationId);

        $clientIpAddr = $donation->donorIp;
        $transaction_id = $queryParams['trans_id'];
        $transactionResult = $this->client->getTransactionResult($transaction_id, $clientIpAddr);

        error_log(print_r($transactionResult), true);

        if ($transactionResult["RESULT_CODE"] == '000') {
            $content = "Donation Completed";
            $donation->status = DonationStatus::COMPLETE();
            $redirectUrl = $queryParams['givewp-success-url'];
        } elseif ($transactionResult['RESULT_CODE'] == "116") {
            $content = "Tranzactie nereusita. Mijloace insuficiente";
            $donation->status = DonationStatus::FAILED();
            $redirectUrl = $queryParams['cancel_url'];
        } else {
            $content = "Tranzactie nereusita. Pentru mai multe detalii rugam sa va adresati bancii emitente a cardului la nr de telefon indicat pe versoul cardului";
            $donation->status = DonationStatus::FAILED();
            $redirectUrl = $queryParams['cancel_url'];
        }
        // Step 3: Use the Donation model to update the donation based on the transaction and response from the gateway.
        // $donation->status = DonationStatus::COMPLETE();
        // $donation->gatewayTransactionId = $gatewayTransactionId;
        $donation->save();

        DonationNote::create([
            'donationId' => $donation->id,
            'content' => $content,//'Donation Completed from MAIB Gateway Offsite.'
        ]);

        // Step 4: Return a RedirectResponse to the GiveWP success page.
        return new RedirectResponse($redirectUrl);
        //return new RedirectResponse($successUrl);
    }

    /**
     * @inerhitDoc
     */
    public function refundDonation(Donation $donation): PaymentRefunded
    {
        // Step 1: refund the donation with your gateway.
        // Step 2: return a command to complete the refund.
        return new PaymentRefunded();
    }
}
`

**Comment by jonwaldstein** [comment]
Hey @cusursuz,

Can you please clarify what you mean by _you can't run handleCreatePaymentRedirect method_, are you receiving an error? If so, please share.

Thanks!

**Reply by cusursuz** [reply]
After redirect I need the method run to verify transaction status.
But it is not running

**Reply by jonwaldstein** [reply]
Can you share what the url looks like when it redirects to your website?

**Reply by cusursuz** [reply]
Url is post,  this is array from POST
Array
(
    [trans_id] => 1RcRYcsmUnlSSLbOLPO8zJ9PWzU=
    [Ucaf_Cardholder_Confirm] => 0
    [cancel_url] => https://wordpress-1146944-4156667.cloudwaysapps.com/sample-page/?giveDonationAction=failedDonation
    [give-gateway-method] => handleCreatePaymentRedirect
    [givewp-gateway-transaction-id] => 1RcRYcsmUnlSSLbOLPO8zJ9PWzU=
    [amount] => 10.00
    [give-route-signature-expiration] => 1707050286
    [item_name] => Donation Form
    [email_address] =donatiinfromsite@gmail.com
    [name_first] => x
    [item_description] => Donation via GiveWP, ID 67
    [notify_url] => https://wordpress-1146944-4156667.cloudwaysapps.com/?give-listener=MAIB
    [give-route-signature-id] => 67
    [m_payment_id] => 67
    [givewp-donation-id] => 67
    [give-route-signature] => d61838c7cbae34c4e44626aaee0e7eb3
    [name_last] => x
    [transaction_id] => 1RcRYcsmUnlSSLbOLPO8zJ9PWzU=
    [return_url] => https://wordpress-1146944-4156667.cloudwaysapps.com?give-listener=give-gateway
    [givewp-success-url] => https://wordpress-1146944-4156667.cloudwaysapps.com/sample-page/?giveDonationAction=showReceipt
    [give-gateway-id] => maib-gateway-offsite
)

**Reply by jonwaldstein** [reply]
@cusursuz it appears your `return_url` is not complete, that url needs to use `$this->generateSecureGatewayRouteUrl()`

**Reply by cusursuz** [reply]
I use same
`$returnUrl = $this->generateSecureGatewayRouteUrl(
'handleCreatePaymentRedirect',
$donation->id,
[
'givewp-donation-id' => $donation->id,
'givewp-success-url' => urlencode(give_get_success_page_uri()),
'givewp-gateway-transaction-id' => urlencode($sms_transaction_id),
]
);`

**Reply by jonwaldstein** [reply]
@cusursuz yeah looks correct, maybe the gateway is stripping the url params on redirect?

**Reply by cusursuz** [reply]
must be POST or GET?

**Reply by jonwaldstein** [reply]
It is intended to be a GET request as the gateway will typically redirect the user back to the website with givewp

**Reply by cusursuz** [reply]
we receive response back with POST request. 
I will ask if bank can give response with GET request

**Reply by cusursuz** [reply]
The bank can only respond using the POST method.
Is there another solution?

**Reply by jonwaldstein** [reply]
@cusursuz that _should_ still be possible, however it looked liked the gateway was still stripping the return_url query params which would mean its not reaching the correct url.  do you have documentation for this gateway you are working with?

**Reply by cusursuz** [reply]
Yes, this is
https://github.com/maibank/maibapi

**Reply by jonwaldstein** [reply]
@cusursuz from the docs it says:

>The second step is the redirected user to the Maib Payment Gateway URL using the TRANSACTION_ID.
When the transaction has been finalised, the Maib Payment Gateway redirects the user to your callback URL where you get the transaction status

I'm not sure how to supply the callback URL - but that should be the url from `$this->generateSecureGatewayRouteUrl()`

**Comment by cusursuz** [comment]
Return url begin with return_url=http://customgateway.local?give-listener=give-gateway&.....
Is it possible to change to ...http://customgateway.local/SOME-PATH?give-listener=give-gateway&.....
Thanks

**Reply by jonwaldstein** [reply]
Out of curiosity, why do you want to change the path?

Technically yes, you could probably change that with php but our generated routes are meant to be used in conjunction with the method specified in `$secureRouteMethods` to work properly with our gateway api.

**Reply by cusursuz** [reply]
I understood you
P.S. It’s more difficult to change the callback at a bank

**Reply by jonwaldstein** [reply]
I saw this in the docs:
>Need to write an email to maib commerce support: [ecom@maib.md](mailto:ecom@maib.md) with the request including The Merchant IP and callback URL, to receive access and certificate for testing.

Do you know if the callback url can be dynamic using an api or does it have to be a specific url including domain name?  That could be a challenge in a WordPress environment where our plugin is distributed. Typically payment gateways will need a specific webhook url to reply asynchronously - but the return url / callback url is usually something given programatically.

**Reply by cusursuz** [reply]
can't be dynamic
We make request to change it for main domain ex.: our_site.md/
